compiler/rustc_infer/src/infer/sub.rs

```diff
@@ -153,6 +153,32 @@ impl<'tcx> TypeRelation<'tcx> for Sub<'_, '_, 'tcx> {
                 );
                 Ok(a)
             }
+            (&ty::GeneratorWitness(a_types), &ty::GeneratorWitness(b_types)) => {
+                // This is worth explaining. GeneratorWitnesses have their
+                // types in binders so that they don't capture any of the
+                // lifetime relations of the generator interiors. However,
+                // the bound lifetimes in these binders aren't really universal,
+                // but instead existential. Also, by construction, two witnesses
+                // are really only going to be compared to each other when
+                // they are from the same generator.
+                //
+                // We use this to simplify the sub and eq relations of generator
+                // interiors significantly, by instead instantiating both
+                // binders with placeholders in the same universe and relating
+                // those directly, instead of using `higher_ranked_sub`, which
+                // in the case of large futures causes many region vids and
+                // universes to be instantiated.
+                assert_eq!(a_types.skip_binder().len(), b_types.skip_binder().len());
+                let universe = infcx.create_next_universe();
+                let a_types =
+                    infcx.replace_bound_vars_with_placeholders_in_universe(a_types, universe);
+                let b_types =
+                    infcx.replace_bound_vars_with_placeholders_in_universe(b_types, universe);
+                for (a, b) in std::iter::zip(a_types, b_types) {
+                    self.relate(a, b)?;
+                }
```

Comment:
that seems really dangerous in cause one of the generators has anonymized bound vars while the other one does not. Anonymization can reorder bound vars which would cause this to fail even though it shouldn't.

I believe we need something like the impl in #97427. I personally would prefer for us to not add the method `replace_bound_vars_with_placeholders_in_universe` as I can't think of any use which does not feel worrying to me.

Reply:
Makes sense. Shouldn't be too hard.